docs/SBVR.md

```diff
@@ -0,0 +1,355 @@
+Rule: It is necessary that each foo has exactly one bar
+```
+
+### Note
+
+A `Note` is a simple uninterpreted comment:
+
+```
+Term: foo
+    Note: this is an arbitrary comment referring to the term 'foo'
+```
+
+### Synonymous Form
+
+The `Synonymous Form` attribute introduces an alternative syntax for a `Fact Type`. As a special case, binary fact types using the `has` verb in the form:
+
+```
+Fact Type: <term-1> has <term-2>
+```
+
+will automatically have an `is of` synonymous form:
+
+```
+Synonymous Form: <term-2> is of <term-1>
+```
+
+Notice that the order of terms is switched in the synonymous form.
+
+### Term Form
+
+For fact types, the `Term Form` attribute introduces a term alias for the fact type. This means we can refer to the fact type anywhere a term is accepted. For example:
+
+```
+Term: foo
+
+Term: bar
+
+Fact Type: foo has bar
+    Term Form: foobar
+
+Fact Type: foobar beyond all recognition
+```
+
+This attribute treats the `foo has bar` join table as the table for the `foobar` term:
+
+```sql
+CREATE TABLE IF NOT EXISTS "foo" (
+        "created at" TIMESTAMP DEFAULT CURRENT_TIMESTAMP NOT NULL
+,       "id" SERIAL NOT NULL PRIMARY KEY
+);
+
+CREATE TABLE IF NOT EXISTS "bar" (
+        "created at" TIMESTAMP DEFAULT CURRENT_TIMESTAMP NOT NULL
+,       "id" SERIAL NOT NULL PRIMARY KEY
+);
+
+CREATE TABLE IF NOT EXISTS "foo-has-bar" (
+        "created at" TIMESTAMP DEFAULT CURRENT_TIMESTAMP NOT NULL
+,       "foo" INTEGER NOT NULL
+,       "bar" INTEGER NOT NULL
+,       "id" SERIAL NOT NULL PRIMARY KEY
+,       "beyond all recognition" INTEGER DEFAULT 0 NOT NULL
+,       FOREIGN KEY ("foo") REFERENCES "foo" ("id")
+,       FOREIGN KEY ("bar") REFERENCES "bar" ("id")
+,       UNIQUE("foo", "bar")
+);
+```
+
+## Constraints
+
+Constraints are english statements interpreted as modal first order logic. Pine accepts two modalities: _alethic_ and _deontic_. The alethic modality is used to state that something _must_ be true and violations are not expected at all. The deontic modality, on the other hand, states that something _should_ be true, but constraint violations are not fatal and potentially recoverable.
+
+To start an alethic statement use `It is necessary that`, followed by the first order constraint. Deontic statements start with `It is obligatory that` instead. Pine also understands the following variations:
+
+| Variation | Equivalent to |
+| --- | --- |
+| `It is impossible that` | It is necessary that not |
+| `It is not possible that` | It is necessary that not |
+| `It is possible that` | It is not necessary that not |
+| `It is prohibited that` | It is obligatory that not |
+| `It is forbidden that` | It is obligatory that not |
+| `It is permitted that` | It is not obligatory that not |
+
+Currently Pine parses modalities but assumes they are all equivalent to necessities. It also assumes that rules that can be represented directly in SQL (schema, UNIQUE/NOT NULL/etc) are alethic, and everything else deontic.
```

Comment:
This is interesting, never fully understood how we differentiate between the two. Can we add an example of a deontic rule here? Also, we should explicitly mention the runtime cost that is associated with the two: as the first comes for "free" since they are enforced by the DB, but we have to explicitly check for other rules after an insert/update. What happens in practice if a deontic rule is violated?

Reply:
Within SBVR, the difference is about what to expect: violations of alethic rules are not expected at all, while violations of deontic rules are not desired but possible. As far as the database is concerned, the difference is how you get an error: either the DB refuses to work (alethic, can't do anything about that unless you change the schema behind Pine's back) or a a validation check fails (deontic, user-driver check that can be ignored). I haven't gone down the code path dealing with validation checks though.

It's hard to talk about runtime costs and the alethic/deontic split in practice because it depends on the DB, the Pine version, and how you phrase your constraint. The basic gist seems to be that if Pine sees an opportunity to encode a constraint into the schema, it takes it. Everything else is put into validation checks.

But regardless, SBVR modalities are completely irrelevant in practice. You could scrap them without losing functionality (even `impossible` is treated as `necessary`), so I don't think it makes much sense to delve into that.